pkg/types/interval.go

```diff
@@ -64,6 +64,61 @@ func (i Interval) Duration() time.Duration {
 	return time.Duration(i.Milliseconds()) * time.Millisecond
 }
 
+// Convert determines the candle open time from a given timestamp
+// eg interval 1 hour and tick at timestamp 00:58 will return timestamp shifted to 00:00
+func (i Interval) Convert(ts time.Time) time.Time {
+	var start = time.Date(ts.Year(), ts.Month(), ts.Day(), ts.Hour(), ts.Second(), 0, 0, ts.Location())
+	switch i {
+	case Interval1s: // default start as defined above
+	case Interval1m:
+		return time.Date(ts.Year(), ts.Month(), ts.Day(), ts.Hour(), ts.Minute(), 0, 0, ts.Location())
+	case Interval3m:
+		return shiftMinute(ts, 3)
+	case Interval5m:
+		return shiftMinute(ts, 5)
+	case Interval15m:
+		return shiftMinute(ts, 15)
+	case Interval30m:
+		return shiftMinute(ts, 30)
+	case Interval1h:
+		return time.Date(ts.Year(), ts.Month(), ts.Day(), ts.Hour(), 0, 0, 0, ts.Location())
+	case Interval2h:
+		return shiftHour(ts, 2)
+	case Interval4h:
+		return shiftHour(ts, 4)
+	case Interval6h:
+		return shiftHour(ts, 6)
+	case Interval12h:
+		return shiftHour(ts, 12)
+	case Interval1d:
+		return time.Date(ts.Year(), ts.Month(), ts.Day(), 0, 0, 0, 0, ts.Location())
+	case Interval3d:
+		return shiftDay(ts, 3)
+	case Interval1w:
+		return shiftDay(ts, 7)
+	case Interval2w:
+		return shiftDay(ts, 14)
+	case Interval1mo:
+		return time.Date(ts.Year(), ts.Month(), 0, 0, 0, 0, 0, ts.Location())
+	}
+	return start
+}
+
+func shiftDay(ts time.Time, shift int) time.Time {
+	day := ts.Day() - (ts.Day() % shift)
+	return time.Date(ts.Year(), ts.Month(), day, 0, 0, 0, 0, ts.Location())
+}
+
+func shiftHour(ts time.Time, shift int) time.Time {
+	hour := ts.Hour() - (ts.Hour() % shift)
+	return time.Date(ts.Year(), ts.Month(), ts.Day(), hour, 0, 0, 0, ts.Location())
```

Comment:
there is a time.Time method called Truncate, you could take a look

Reply:
I dont think this is gonna work in this case:

`Truncate operates on the time as an absolute duration since the zero time; it does not operate on the presentation form of the time. Thus, Truncate(Hour) may return a time with a non-zero minute, depending on the time's Location.`

Reply:
and alternatively if ts is 00:34:27 you want to shift back to 30 not truncate to zero

Reply:
I tried but didn't get the test to pass with shifting

```
func shiftHour(ts time.Time, shift int) time.Time {
	hour := time.Duration(ts.Hour()) * time.Hour
	return ts.Truncate(hour).Add(shift* time.Hour)
}
```